{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a71a3685-b41b-4dd7-8987-fb26729209a6",
   "metadata": {},
   "source": [
    "# DeepASM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30ca0fb9-e45a-4d26-be20-575214beec95",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Install packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "7aa06369-0a4f-4120-8b0f-a4dfcb5e7627",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To manipulate HDF5 files (RUN FOR ALL MODELS)\n",
    "!pip3 install --upgrade tables"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ecb353-258f-451d-8c2a-2488137356b1",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9eb55e93-2391-4791-bca2-a69b4ec7a4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "# Python packages for data, stats, and visualization\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns \n",
    "\n",
    "# Machine learning libraries\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler\n",
    "import tensorflow as tf\n",
    "from tensorflow import keras\n",
    "#from tensorflow.keras.models import load_model\n",
    "from tensorflow.keras import layers\n",
    "import sklearn\n",
    "from sklearn import metrics\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Kernel functions\n",
    "from sklearn.neighbors import KernelDensity\n",
    "from numpy import asarray\n",
    "from matplotlib import pyplot\n",
    "from numpy import exp\n",
    "\n",
    "# Dimensionality reduction\n",
    "from sklearn.decomposition import PCA, KernelPCA, NMF, TruncatedSVD\n",
    "from sklearn.manifold import TSNE, LocallyLinearEmbedding, SpectralEmbedding\n",
    "\n",
    "# To get the time\n",
    "from datetime import datetime\n",
    "\n",
    " \n",
    "# Figure parameters\n",
    "mpl.rcParams['figure.figsize'] = (10, 10)\n",
    "mpl.rcParams['axes.titlesize'] = 15\n",
    "mpl.rcParams['axes.labelsize'] = 12\n",
    "mpl.rcParams['xtick.labelsize'] = 12\n",
    "mpl.rcParams['ytick.labelsize'] = 12\n",
    "colors = plt.rcParams['axes.prop_cycle'].by_key()['color']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "683ebf71-130f-4ea0-8eaa-c27478ebb990",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:53) \n",
      "[GCC 9.4.0]\n",
      "TensorFlow version: 2.7.0\n",
      "Keras version: 2.7.0\n",
      "Numpy version: 1.19.5\n",
      "Num GPUs Available:  1\n"
     ]
    }
   ],
   "source": [
    "# Print different versions\n",
    "print(sys.version)\n",
    "print(\"TensorFlow version:\", tf.__version__)\n",
    "print(\"Keras version:\", keras.__version__)\n",
    "print(\"Numpy version:\", np.__version__)\n",
    "print(\"Num GPUs Available: \", len(tf.config.list_physical_devices('GPU')))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abddb628-6484-4f52-9386-ced2fb0f456a",
   "metadata": {},
   "source": [
    "## GCP Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2cb33deb-7b19-4795-acd3-4511eb464de1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import raw data from bucket. False if you want to import the processed dataset\n",
    "IMPORT_RAW_FROM_BUCKET = True\n",
    "\n",
    "# Export data after it's been prepared\n",
    "EXPORT_PROCESSED_DATA = True\n",
    "\n",
    "# Bucket name where the training datasets are\n",
    "DEEPASM_BUCKET=\"deepasm\"\n",
    "\n",
    "# GCP variable\n",
    "import os\n",
    "projectid = 'hackensack-tyco'\n",
    "os.environ[\"GOOGLE_CLOUD_PROJECT\"] = 'hackensack-tyco'\n",
    "cloud_bucket = \"gs://deepasm/colab\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abb144c7-b3d9-4ca7-836b-027939589d48",
   "metadata": {},
   "source": [
    "## Model variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d49f672e-93d9-485a-9914-010650158ce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MODELS FOR WHICH WE NEED TO RECORD THE RESULTS\n",
    "\n",
    "models = ['linear', 'perceptron', 'simple_cnn',  'cnn', 'simple_rnn', 'rnn']\n",
    "#models = ['simple_rnn']\n",
    "\n",
    "# Loss is better than AUC for monitoring\n",
    "PARAM_TO_CHANGE = \"keep_chr\"\n",
    "\n",
    "#--------------------------------------------------\n",
    "# Parameters common to all models\n",
    "\n",
    "# Number of rows to take into the dataset after import\n",
    "NB_ROWS_RAW_DATASET = int(1e3) # The maximum is 5e6. We use 200k to test the code\n",
    "\n",
    "# Minimum correlation factor\n",
    "MIN_CORR = 0.03\n",
    "\n",
    "# Size of the genomic window\n",
    "GENOMIC_INTERVAL = 1000\n",
    "\n",
    "# Kernel values for probability estimates\n",
    "KERNEL_FM_NB_VALUES = 10\n",
    "KERNEL_FM_BANDWIDTH = 0.1\n",
    "KERNEL_COV_NB_MAX = 200\n",
    "KERNEL_COV_NB_STEP = 40\n",
    "KERNEL_COV_BANDWIDTH = 20\n",
    "\n",
    "# Normalization method\n",
    "norm_method = \"z_score\" # \"min_max\" or \"z_score\"\n",
    "\n",
    "# Keep the chromosomes in the model (correlate poorly with ASM)\n",
    "KEEP_CHR_IN_MODEL = False\n",
    "\n",
    "# Early stopping\n",
    "EARLY_STOPPING = tf.keras.callbacks.EarlyStopping(\n",
    "    monitor='val_loss', \n",
    "    verbose=0,\n",
    "    patience=5,\n",
    "    mode='auto',\n",
    "    restore_best_weights=True)\n",
    "\n",
    "# Percentage of data points to be used in the Test dataset\n",
    "TEST_SPLIT = 0.2\n",
    "\n",
    "# Percentage of datapoints used between training and validation\n",
    "VALIDATION_SPLIT = 0.3 # How to divide the training dataset for validation\n",
    "\n",
    "EPOCHS = 100 # We have so many datapoints that 20 epochs are enough to stabilize the training\n",
    "BATCH_SIZE = 1000 # to get a few identified ASM we need at a few hundreds since the\n",
    "# frequency of ASM is 1.38%\n",
    "# A batch size of 1000 will run into a memory error on TF 2.7\n",
    "\n",
    "# Regularlization L1 and L2 (defaults are l1 = 0.01 and l2 = 0.01)\n",
    "L1_R = 0\n",
    "L2_R = 1e-3\n",
    "\n",
    "#--------------------------------------------------\n",
    "# Parameters common to neural network models\n",
    "ACTIVATION_FUNCTION = 'relu' # 'tanh' # or 'relu' or 'gelu (Gaussian Error Linear Unit)'\n",
    "NB_NODES_PERCEPTRON = 10\n",
    "NB_LAYERS_PERCEPTRON = 5\n",
    "NB_NODES_AFTER_CNN = 2\n",
    "CNN_FILTERS = 8\n",
    "CNN_KERNEL = 100 # Must be smaller than the genomic region (250). The av distance between CpG is 37 bp and the std dev of the distances between cpgs is 24 bp\n",
    "LEARNING_RATE = 3e-4 \n",
    "\n",
    "# Learning rate was taken from this\n",
    "# http://karpathy.github.io/2019/04/25/recipe/#2-set-up-the-end-to-end-trainingevaluation-skeleton--get-dumb-baselines\n",
    "\n",
    "#--------------------------------------------------\n",
    "# Parameters common to RNN\n",
    "\n",
    "RNN_UNITS = 64 # 64 orginally\n",
    "\n",
    "#--------------------------------------------------\n",
    "# SPECIFIC TO RANDOM FOREST ALGORITHM\n",
    "use_raw_df_for_forest_models = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c22515ca-30c7-4a83-a982-6eb67b884f37",
   "metadata": {},
   "source": [
    "## ML evaluation metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2f38b202-fe69-489c-aae0-638bcd40b2d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-03-08 18:25:48.975753: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2022-03-08 18:25:49.586203: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38444 MB memory:  -> device: 0, name: A100-SXM4-40GB, pci bus id: 0000:00:04.0, compute capability: 8.0\n"
     ]
    }
   ],
   "source": [
    "METRICS = [\n",
    "      keras.metrics.TruePositives(name='tp'),\n",
    "      keras.metrics.FalsePositives(name='fp'),\n",
    "      keras.metrics.TrueNegatives(name='tn'),\n",
    "      keras.metrics.FalseNegatives(name='fn'), \n",
    "      keras.metrics.BinaryAccuracy(name='accuracy'),\n",
    "      keras.metrics.Precision(name='precision'),\n",
    "      keras.metrics.Recall(name='sensitivity'),\n",
    "      keras.metrics.AUC(name='auc')\n",
    "      ]\n",
    "\n",
    "def plot_metrics(history):\n",
    "  metrics =  ['loss', 'auc', 'precision', 'sensitivity']\n",
    "  plt.figure(figsize=(10,10))\n",
    "  for n, metric in enumerate(metrics):\n",
    "    name = metric.replace(\"_\",\" \").capitalize()\n",
    "    plt.subplot(2,2,n+1)\n",
    "    plt.plot(history.epoch,  history.history[metric], color=colors[0], label='Train')\n",
    "    plt.plot(history.epoch, history.history['val_'+metric],\n",
    "             color=colors[0], linestyle=\"--\", label='Val')\n",
    "    plt.xlabel('Epoch')\n",
    "    plt.ylabel(name)\n",
    "    plt.ylim([0,1])\n",
    "\n",
    "    plt.legend()\n",
    "\n",
    "\n",
    "def display_results(df_results):\n",
    "  print(\"Loss\", np.round(df_results[0], 3))\n",
    "  print(\"True positives\", np.round(df_results[1], 3))\n",
    "  print(\"False positives\", np.round(df_results[2], 3))\n",
    "  print(\"True negatives\", np.round(df_results[3], 3))\n",
    "  print(\"False negatives\", np.round(df_results[4], 3))\n",
    "  print(\"Accuracy\", np.round(df_results[5], 3))\n",
    "  print(\"Precision\", np.round(df_results[6], 3))\n",
    "  print(\"Sensitivity\", np.round(df_results[7], 3))\n",
    "  print(\"AUC\", np.round(df_results[8], 3))\n",
    "\n",
    "def plot_roc(name, labels, predictions, **kwargs):\n",
    "  fp, tp, _ = sklearn.metrics.roc_curve(labels, predictions)\n",
    "\n",
    "  plt.plot(100*fp, 100*tp, label=name, linewidth=2, **kwargs)\n",
    "  plt.xlabel('False positives [%]')\n",
    "  plt.ylabel('True positives [%]')\n",
    "  plt.xlim([-0.5,80])\n",
    "  plt.ylim([0,100.5])\n",
    "  plt.grid(True)\n",
    "  ax = plt.gca()\n",
    "  ax.set_aspect('equal')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "025e5515-6785-4461-b139-ced8487e2eee",
   "metadata": {},
   "source": [
    "## Import raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d440a700-3f8d-443b-87f4-3036f2a0f5b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files to download: 21\n",
      "Appending file...\n"
     ]
    }
   ],
   "source": [
    "if IMPORT_RAW_FROM_BUCKET == True:\n",
    "    !gsutil ls gs://$DEEPASM_BUCKET/$GENOMIC_INTERVAL*bp/encode_training_data/*.json > list_to_download.txt\n",
    "    files_to_download_df = pd.read_csv('list_to_download.txt', header=None)\n",
    "    print(\"Number of files to download:\", files_to_download_df.shape[0])\n",
    "\n",
    "    imported_df = pd.DataFrame()\n",
    "\n",
    "    for index_file in range(range(files_to_download_df.shape[0])): #range(files_to_download_df.shape[0])\n",
    "        file_name_bucket = files_to_download_df[0][index_file]\n",
    "        local_file_name = \"training_\" + str(index_file) + \".json\"\n",
    "        \n",
    "        # Download the file from bucket\n",
    "        !gsutil cp $file_name_bucket $local_file_name\n",
    "        \n",
    "        print(\"Appending file...\")\n",
    "        imported_df = imported_df.append(pd.read_json(local_file_name, lines = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f83a257f-e990-4f89-bec0-0011061db225",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of the imported dataset: (214876, 18)\n"
     ]
    }
   ],
   "source": [
    "print(\"Size of the imported dataset:\", imported_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd44574e-a6e1-4e5b-b254-bcf6959d60a0",
   "metadata": {},
   "source": [
    "## Prepare the features\n",
    "\n",
    "Note: we do not randomize the rows because the scripts preceding this notebook already sampled the rows."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b3f096d-c8fd-4270-b822-97fd20c94331",
   "metadata": {},
   "source": [
    "### Copy & clean dataframe "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0b1bbac4-bc09-4786-b688-b8976eeac8bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy of the dataframe\n",
    "raw_df = imported_df.copy()\n",
    "raw_df = raw_df.head(NB_ROWS_RAW_DATASET)\n",
    "\n",
    "# We remove the chromosomes X and Y (no ASM)\n",
    "raw_df = raw_df.loc[raw_df['chr'] != 'X']\n",
    "raw_df = raw_df.loc[raw_df['chr'] != 'Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "26e8e447-ea95-4b96-a467-19606d3f0f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Size of the dataset: \", raw_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8a0c078-541c-4563-a68a-4dd87bb34e5f",
   "metadata": {},
   "source": [
    "### Calculate the distance between CpGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1f2362ad-6e00-4d94-961d-8081ada1d11b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a function to calculate the distance between CpGs (~3min)\n",
    "def dist_cpg(cpg_pos):\n",
    "  distances = []\n",
    "  for index in range(len(cpg_pos)):\n",
    "    if index >= len(cpg_pos)-1:\n",
    "      return distances\n",
    "    else:\n",
    "      distances.append(cpg_pos[index + 1] - cpg_pos[index])\n",
    "  return distances\n",
    "\n",
    "# Apply the function \"distance\" to the array of CpG positions\n",
    "raw_df['cpg_dist'] = raw_df['cpg_pos'].apply(lambda x: dist_cpg(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "fe7a2cbb-37a8-4e15-b420-e61802e539c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm_snp</th>\n",
       "      <th>sample_category</th>\n",
       "      <th>chr</th>\n",
       "      <th>region_inf</th>\n",
       "      <th>region_sup</th>\n",
       "      <th>region_nb_cpg</th>\n",
       "      <th>nb_cpg_found</th>\n",
       "      <th>nb_reads</th>\n",
       "      <th>dnase</th>\n",
       "      <th>encode_ChiP_V2</th>\n",
       "      <th>tf_motifs</th>\n",
       "      <th>global_cpg_fm</th>\n",
       "      <th>tot_nb_cpg</th>\n",
       "      <th>tot_nb_reads</th>\n",
       "      <th>read_fm</th>\n",
       "      <th>cpg_fm</th>\n",
       "      <th>cpg_cov</th>\n",
       "      <th>cpg_pos</th>\n",
       "      <th>cpg_dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>167184001</td>\n",
       "      <td>167185000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.815</td>\n",
       "      <td>21223988</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[0, 0, 0, 1, 0, 0.5, 0, 1, 0, 1, 0.5, 0.333, 1...</td>\n",
       "      <td>[0.742, 0.788, 0.25]</td>\n",
       "      <td>[31, 33, 40]</td>\n",
       "      <td>[167184497, 167184558, 167184725]</td>\n",
       "      <td>[61, 167]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>166743001</td>\n",
       "      <td>166744000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.815</td>\n",
       "      <td>21223988</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.5, 1, 1...</td>\n",
       "      <td>[1, 0.9500000000000001, 0.929]</td>\n",
       "      <td>[38, 40, 42]</td>\n",
       "      <td>[166743591, 166743715, 166743733]</td>\n",
       "      <td>[124, 18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>22376001</td>\n",
       "      <td>22377000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.815</td>\n",
       "      <td>21223988</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.5, 1, 0...</td>\n",
       "      <td>[0.783, 0.838, 0.765]</td>\n",
       "      <td>[46, 37, 34]</td>\n",
       "      <td>[22376280, 22376368, 22376447]</td>\n",
       "      <td>[88, 79]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>63617001</td>\n",
       "      <td>63618000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.815</td>\n",
       "      <td>21223988</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[1, 0, 1, 1, 1, 0.667, 1, 0.667, 1, 1, 1, 1, 0...</td>\n",
       "      <td>[0.636, 0.9570000000000001, 0.979]</td>\n",
       "      <td>[44, 47, 47]</td>\n",
       "      <td>[63617746, 63617751, 63617797]</td>\n",
       "      <td>[5, 46]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>188785001</td>\n",
       "      <td>188786000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>144</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.815</td>\n",
       "      <td>21223988</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[1, 0.5, 0, 0, 1, 0.5, 0, 0, 0, 0.5, 1, 0, 0, ...</td>\n",
       "      <td>[0.543, 0.356, 0.646]</td>\n",
       "      <td>[46, 90, 96]</td>\n",
       "      <td>[188785538, 188785705, 188785753]</td>\n",
       "      <td>[167, 48]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>113881001</td>\n",
       "      <td>113882000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.786</td>\n",
       "      <td>17213478</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[1, 1, 1, 1, 0.667, 0.667, 1, 1, 1, 1, 1, 0.66...</td>\n",
       "      <td>[0.97, 0.765, 0.8290000000000001]</td>\n",
       "      <td>[33, 34, 35]</td>\n",
       "      <td>[113881397, 113881438, 113881453]</td>\n",
       "      <td>[41, 15]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>166743001</td>\n",
       "      <td>166744000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.786</td>\n",
       "      <td>17213478</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 0.5, 1, 1, 1, 1, 1, 1...</td>\n",
       "      <td>[0.919, 0.962, 0.9490000000000001]</td>\n",
       "      <td>[37, 26, 39]</td>\n",
       "      <td>[166743591, 166743715, 166743733]</td>\n",
       "      <td>[124, 18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>83870001</td>\n",
       "      <td>83871000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.786</td>\n",
       "      <td>17213478</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 0.667, 1, 1, 1, 1, 0....</td>\n",
       "      <td>[0.9520000000000001, 0.857, 0.909]</td>\n",
       "      <td>[21, 28, 22]</td>\n",
       "      <td>[83870175, 83870190, 83870236]</td>\n",
       "      <td>[15, 46]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>90796001</td>\n",
       "      <td>90797000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.786</td>\n",
       "      <td>17213478</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.667, 1, 1,...</td>\n",
       "      <td>[0.962, 0.722, 0.92]</td>\n",
       "      <td>[26, 36, 25]</td>\n",
       "      <td>[90796083, 90796160, 90796178]</td>\n",
       "      <td>[77, 18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>121825001</td>\n",
       "      <td>121826000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.786</td>\n",
       "      <td>17213478</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[1, 0.333, 1, 1, 1, 1, 0, 1, 0, 0, 0.333, 0.33...</td>\n",
       "      <td>[0.788, 0.774, 0.742]</td>\n",
       "      <td>[33, 31, 31]</td>\n",
       "      <td>[121825759, 121825766, 121825777]</td>\n",
       "      <td>[7, 11]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>985 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     asm_snp  sample_category chr  region_inf  region_sup  region_nb_cpg  \\\n",
       "0          0                0   5   167184001   167185000              3   \n",
       "1          0                0   5   166743001   166744000              3   \n",
       "2          0                0   6    22376001    22377000              3   \n",
       "3          0                0  14    63617001    63618000              3   \n",
       "4          0                0   2   188785001   188786000              3   \n",
       "..       ...              ...  ..         ...         ...            ...   \n",
       "995        0                0   3   113881001   113882000              3   \n",
       "996        0                0   5   166743001   166744000              3   \n",
       "997        0                0   1    83870001    83871000              3   \n",
       "998        0                0   1    90796001    90797000              3   \n",
       "999        0                0  10   121825001   121826000              3   \n",
       "\n",
       "     nb_cpg_found  nb_reads  dnase  encode_ChiP_V2  tf_motifs  global_cpg_fm  \\\n",
       "0               3        71      1               0          0          0.815   \n",
       "1               3        69      0               0          2          0.815   \n",
       "2               3        75      0               0          0          0.815   \n",
       "3               3        57      1               0          3          0.815   \n",
       "4               3       144      2               1          0          0.815   \n",
       "..            ...       ...    ...             ...        ...            ...   \n",
       "995             3        42      1               0          8          0.786   \n",
       "996             3        60      0               0          2          0.786   \n",
       "997             3        32      1               0          6          0.786   \n",
       "998             3        49      0               0          0          0.786   \n",
       "999             3        36      0               0          0          0.786   \n",
       "\n",
       "     tot_nb_cpg  tot_nb_reads  \\\n",
       "0      21223988     307561222   \n",
       "1      21223988     307561222   \n",
       "2      21223988     307561222   \n",
       "3      21223988     307561222   \n",
       "4      21223988     307561222   \n",
       "..          ...           ...   \n",
       "995    17213478     222365425   \n",
       "996    17213478     222365425   \n",
       "997    17213478     222365425   \n",
       "998    17213478     222365425   \n",
       "999    17213478     222365425   \n",
       "\n",
       "                                               read_fm  \\\n",
       "0    [0, 0, 0, 1, 0, 0.5, 0, 1, 0, 1, 0.5, 0.333, 1...   \n",
       "1    [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.5, 1, 1...   \n",
       "2    [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.5, 1, 0...   \n",
       "3    [1, 0, 1, 1, 1, 0.667, 1, 0.667, 1, 1, 1, 1, 0...   \n",
       "4    [1, 0.5, 0, 0, 1, 0.5, 0, 0, 0, 0.5, 1, 0, 0, ...   \n",
       "..                                                 ...   \n",
       "995  [1, 1, 1, 1, 0.667, 0.667, 1, 1, 1, 1, 1, 0.66...   \n",
       "996  [0, 1, 1, 1, 1, 1, 1, 1, 0.5, 1, 1, 1, 1, 1, 1...   \n",
       "997  [1, 1, 1, 1, 1, 1, 1, 1, 0.667, 1, 1, 1, 1, 0....   \n",
       "998  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.667, 1, 1,...   \n",
       "999  [1, 0.333, 1, 1, 1, 1, 0, 1, 0, 0, 0.333, 0.33...   \n",
       "\n",
       "                                 cpg_fm       cpg_cov  \\\n",
       "0                  [0.742, 0.788, 0.25]  [31, 33, 40]   \n",
       "1        [1, 0.9500000000000001, 0.929]  [38, 40, 42]   \n",
       "2                 [0.783, 0.838, 0.765]  [46, 37, 34]   \n",
       "3    [0.636, 0.9570000000000001, 0.979]  [44, 47, 47]   \n",
       "4                 [0.543, 0.356, 0.646]  [46, 90, 96]   \n",
       "..                                  ...           ...   \n",
       "995   [0.97, 0.765, 0.8290000000000001]  [33, 34, 35]   \n",
       "996  [0.919, 0.962, 0.9490000000000001]  [37, 26, 39]   \n",
       "997  [0.9520000000000001, 0.857, 0.909]  [21, 28, 22]   \n",
       "998                [0.962, 0.722, 0.92]  [26, 36, 25]   \n",
       "999               [0.788, 0.774, 0.742]  [33, 31, 31]   \n",
       "\n",
       "                               cpg_pos   cpg_dist  \n",
       "0    [167184497, 167184558, 167184725]  [61, 167]  \n",
       "1    [166743591, 166743715, 166743733]  [124, 18]  \n",
       "2       [22376280, 22376368, 22376447]   [88, 79]  \n",
       "3       [63617746, 63617751, 63617797]    [5, 46]  \n",
       "4    [188785538, 188785705, 188785753]  [167, 48]  \n",
       "..                                 ...        ...  \n",
       "995  [113881397, 113881438, 113881453]   [41, 15]  \n",
       "996  [166743591, 166743715, 166743733]  [124, 18]  \n",
       "997     [83870175, 83870190, 83870236]   [15, 46]  \n",
       "998     [90796083, 90796160, 90796178]   [77, 18]  \n",
       "999  [121825759, 121825766, 121825777]    [7, 11]  \n",
       "\n",
       "[985 rows x 19 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ebc6f7c-ccef-4681-a609-dbde5ac5e379",
   "metadata": {},
   "source": [
    "### Convert arrays into numerical features\n",
    "\n",
    "To do this, we use kernel estimates as well as simpler metrics like mean and standard deviation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6653f9e8-b3a9-4077-8bd9-52d25fd22626",
   "metadata": {},
   "source": [
    "#### Kernel functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a5452c4c-7ff6-43a5-97ec-da1ccbdee6b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.1156, 0.0702, 0.0169, 0.0169, 0.0702, 0.1156, 0.0714, 0.0574,\n",
       "       0.5099, 2.2795, 3.7582])"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FRACTIONAL METHYLATION\n",
    "\n",
    "# Values for fractional methylation (between 0 and 1)\n",
    "values_for_kernel_fm = asarray([value for value in range(0, KERNEL_FM_NB_VALUES+1)])\n",
    "values_for_kernel_fm = values_for_kernel_fm / KERNEL_FM_NB_VALUES\n",
    "print(\"X-axis values used for the FM kernel estimate:\", values_for_kernel_fm)\n",
    "values_for_kernel_fm = values_for_kernel_fm.reshape((len(values_for_kernel_fm), 1))\n",
    "\n",
    "# Build Kernel model\n",
    "kernel_fm_model = KernelDensity(bandwidth=KERNEL_FM_BANDWIDTH, kernel='gaussian')\n",
    "\n",
    "# Function to be applied to each array in the columns read_fm and cpg_fm\n",
    "def estimate_kernels_fm(x):\n",
    "  sample = np.reshape(x, (len(x), 1))\n",
    "  kernel_fm_model.fit(sample)\n",
    "  probabilities = kernel_fm_model.score_samples(values_for_kernel_fm)\n",
    "  probabilities = exp(probabilities)\n",
    "  return np.round(probabilities, 4)\n",
    "\n",
    "# Try function\n",
    "estimate_kernels_fm(raw_df['read_fm'][1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "97c6d384-d6a5-4ad9-bc40-74584337290f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.    , 1.3298, 0.    , 0.    , 0.    ])"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# COVERAGE AND CPG DISTANCE\n",
    "\n",
    "# Values for fractional methylation (between 0 and 1)\n",
    "values_for_kernel_cov = asarray([value for value in range(0, KERNEL_COV_NB_MAX, KERNEL_COV_NB_STEP)])\n",
    "print(\"Values used in kernel estimate:\", values_for_kernel_cov)\n",
    "values_for_kernel_cov = values_for_kernel_cov.reshape((len(values_for_kernel_cov), 1))\n",
    "\n",
    "# Build Kernel model\n",
    "kernel_cov_model = KernelDensity(bandwidth=KERNEL_COV_BANDWIDTH, kernel='gaussian')\n",
    "\n",
    "# Function to be applied to each array in the columns read_fm and cpg_fm\n",
    "def estimate_kernels_cov(x):\n",
    "  sample = np.reshape(x, (len(x), 1))\n",
    "  kernel_fm_model.fit(sample)\n",
    "  probabilities = kernel_fm_model.score_samples(values_for_kernel_cov)\n",
    "  probabilities = exp(probabilities)\n",
    "  return np.round(probabilities, 4)\n",
    "\n",
    "# Try function\n",
    "estimate_kernels_cov(raw_df['cpg_cov'][1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ed3b36-3216-458e-83f5-4e24c38c814e",
   "metadata": {},
   "source": [
    "#### Test kernel estimates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8775a79c-f61b-48fb-bd6e-555c3d12e478",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Cannot take a larger sample than population when 'replace=False'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_22334/4220519270.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mn_extract\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mextract_asm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mraw_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mraw_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'asm_snp'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_extract\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mignore_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mextract_noasm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mraw_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mraw_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'asm_snp'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_extract\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mignore_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mn_x\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mround\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msqrt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_extract\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36msample\u001b[0;34m(self, n, frac, replace, weights, random_state, axis, ignore_index)\u001b[0m\n\u001b[1;32m   5363\u001b[0m             )\n\u001b[1;32m   5364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5365\u001b[0;31m         \u001b[0mlocs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchoice\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis_length\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mweights\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5366\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlocs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5367\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mignore_index\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mmtrand.pyx\u001b[0m in \u001b[0;36mnumpy.random.mtrand.RandomState.choice\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot take a larger sample than population when 'replace=False'"
     ]
    }
   ],
   "source": [
    "variable_to_plot = 'cpg_fm' # cpg_fm or read_fm or cpg_dist or cpg_cov\n",
    "\n",
    "n_extract = 10\n",
    "extract_asm = raw_df[raw_df['asm_snp'] == 1].sample(n=n_extract, ignore_index = True)\n",
    "extract_noasm = raw_df[raw_df['asm_snp'] == 0].sample(n=n_extract, ignore_index = True)\n",
    "n_x = round(np.sqrt(n_extract))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6eb3ee5-1333-43a5-99f0-a7481f682a7d",
   "metadata": {},
   "source": [
    "##### Plots for regions with ASM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c309fc8e-8acb-4955-bb8e-53fa1b602fd8",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'n_x' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_22334/78189908.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrcParams\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'figure.figsize'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_x\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_x\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msharey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msharex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtight_layout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_x\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m   \u001b[0;32mfor\u001b[0m \u001b[0mm\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_x\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'n_x' is not defined"
     ]
    }
   ],
   "source": [
    "mpl.rcParams['figure.figsize'] = (10, 10)\n",
    "fig, axs = plt.subplots(n_x, n_x, sharey=True, sharex=True, tight_layout=True)\n",
    "\n",
    "for k in range(n_x):\n",
    "  for m in range(n_x):\n",
    "\n",
    "    # Print data distribution\n",
    "    data_distribution = extract_asm[variable_to_plot][k+m]\n",
    "    axs[k,m].hist(data_distribution, density = True, bins = 10)\n",
    "\n",
    "    # Print kernel density\n",
    "    if 'fm' in variable_to_plot:\n",
    "        #print(\"Using the FM kernel estimates\")\n",
    "        kernel_probabilities = estimate_kernels_fm(data_distribution)\n",
    "        values = values_for_kernel_fm\n",
    "    else:\n",
    "        #print(\"Using the COV kernel estimates\")\n",
    "        kernel_probabilities = estimate_kernels_cov(data_distribution)\n",
    "        values = values_for_kernel_cov\n",
    "    axs[k,m].plot(values[:], kernel_probabilities)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23eef467-301c-4c3a-8c14-65c63bd6704e",
   "metadata": {},
   "source": [
    "##### Plots for regions without ASM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "305fe19d-1890-45e8-9429-bd26f9503dbd",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'n_x' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_22334/1521153968.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrcParams\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'figure.figsize'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_x\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_x\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msharey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msharex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtight_layout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_x\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m   \u001b[0;32mfor\u001b[0m \u001b[0mm\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_x\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'n_x' is not defined"
     ]
    }
   ],
   "source": [
    "mpl.rcParams['figure.figsize'] = (10, 10)\n",
    "fig, axs = plt.subplots(n_x, n_x, sharey=True, sharex=True, tight_layout=True)\n",
    "\n",
    "for k in range(n_x):\n",
    "  for m in range(n_x):\n",
    "\n",
    "    # Print data distribution\n",
    "    data_distribution = extract_noasm[variable_to_plot][k+m]\n",
    "    axs[k,m].hist(data_distribution, density = True, bins = 10)\n",
    "\n",
    "    # Print kernel density\n",
    "    if 'fm' in variable_to_plot:\n",
    "        #print(\"Using the FM kernel estimates\")\n",
    "        kernel_probabilities = estimate_kernels_fm(data_distribution)\n",
    "        values = values_for_kernel_fm\n",
    "    else:\n",
    "        #print(\"Using the COV kernel estimates\")\n",
    "        kernel_probabilities = estimate_kernels_cov(data_distribution)\n",
    "        values = values_for_kernel_cov\n",
    "    axs[k,m].plot(values[:], kernel_probabilities)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8474e90d-b2cf-4b02-9627-ebb593bc1a2b",
   "metadata": {},
   "source": [
    "#### Calculate the mean, std, and kernel estimates of arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6b85aea5-083a-47ca-93cc-7a554a3f7cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_arrays(df, column_name):\n",
    "  \"\"\"Inputs: dataframe and a column name that contains arrays\"\"\"\n",
    "\n",
    "  # Mean and Standard deviation\n",
    "  std_name = \"std_\" + column_name\n",
    "  av_name = \"mean_\" + column_name\n",
    "\n",
    "  print(\"Calculating the standard deviation\")\n",
    "  df[std_name] = df[column_name].apply(lambda x: np.round(np.std(x), 4))\n",
    "  print(\"Calculating the average\")\n",
    "  df[av_name] = df[column_name].apply(lambda x: np.round(np.mean(x), 4))\n",
    "  \n",
    "  # Kernel density estimates\n",
    "  kernel_name = \"kernel_\" + column_name\n",
    "  if (column_name == 'cpg_cov' or column_name == 'cpg_dist'):\n",
    "    print(\"Calculating the proba distribution for cov or dist\")\n",
    "    df[kernel_name] = df[column_name].apply(lambda x: estimate_kernels_cov(x))\n",
    "  else:\n",
    "    print(\"Calculating the proba distribution for fractional methylation\")\n",
    "    df[kernel_name] = df[column_name].apply(lambda x: estimate_kernels_fm(x))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "da7d5288-af28-40f5-8bb5-9abf3d30014c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function\n",
    "for col in ['read_fm', 'cpg_fm', 'cpg_cov', 'cpg_dist']:\n",
    "  print(\"Column: \", col)\n",
    "  convert_arrays(raw_df, col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "3522320d-cfe7-457d-9d50-e6394d71a68a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_kernel_array(col):\n",
    "    # Col must be a column of kernel estimate arrays\n",
    "    print(\"Processing:\", col)\n",
    "    kernel_name_list = []\n",
    "    if 'fm' in col:\n",
    "        values = values_for_kernel_fm\n",
    "    else:\n",
    "        values = values_for_kernel_cov\n",
    "    # Create a list of the new column names\n",
    "    for k in range(0, values.shape[0]):\n",
    "        kernel_name = col + \"_kernel_\" + str(k)\n",
    "        kernel_name_list = kernel_name_list + [kernel_name]\n",
    "    print(kernel_name_list)\n",
    "    \n",
    "    # Create the additional columns\n",
    "    kernel_estimates_column = \"kernel_\" + col\n",
    "    raw_df[kernel_name_list] = pd.DataFrame(raw_df[kernel_estimates_column].tolist(), index= raw_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2052717c-c317-4b71-bce8-fe5729896820",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the function\n",
    "for col in ['read_fm', 'cpg_fm', 'cpg_cov', 'cpg_dist']:\n",
    "            export_kernel_array(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "3889362e-ba69-4a65-8de1-92be70ff8c06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete columns that we no longer need\n",
    "for col in ['read_fm', 'cpg_dist', 'kernel_cpg_cov', \n",
    "            'kernel_cpg_dist', 'kernel_cpg_fm', 'kernel_read_fm']:\n",
    "    raw_df.drop(col, axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6fb4754-b931-4004-949b-80652daa233a",
   "metadata": {},
   "source": [
    "### Convert epigenetic signals into dummy variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "b79f67df-b84f-4456-99a7-93e12d2916ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_epi_signal(epi_signal):\n",
    "  print(\"Processing signal\", epi_signal)\n",
    "  unique_values = raw_df[epi_signal].unique()\n",
    "  print(unique_values)\n",
    "  min_epi_value = 0 # It's always zero (no signal) for all signals\n",
    "  median_epi_value = np.median(unique_values)\n",
    "  print(\"Median epi value:\", median_epi_value)\n",
    "  epi_signal_null = epi_signal + \"_null\"\n",
    "  epi_signal_low = epi_signal + \"_low\"\n",
    "  epi_signal_high = epi_signal + \"_high\"\n",
    "  raw_df[epi_signal_null] = raw_df[epi_signal].apply(lambda x: 1 if x == min_epi_value else 0)\n",
    "  raw_df[epi_signal_low] = raw_df[epi_signal].apply(lambda x: 1 if (x > min_epi_value and x <= median_epi_value) else 0)\n",
    "  raw_df[epi_signal_high] = raw_df[epi_signal].apply(lambda x: 1 if x > median_epi_value else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "e43bdee0-cd0f-46fd-b54a-301b1b3975cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to all epigenetic signals\n",
    "for epi_signal in ['dnase', 'encode_ChiP_V2', 'tf_motifs']:\n",
    "  convert_epi_signal(epi_signal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "10fceac6-13e4-44f9-80e7-cad08a4f7ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete the raw epigenetic signals\n",
    "for epi_signal in ['dnase', 'encode_ChiP_V2', 'tf_motifs']:\n",
    "  raw_df.drop(epi_signal, axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "9229e490-b543-4f31-b69f-70cc5b3e8a4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['asm_snp', 'sample_category', 'chr', 'region_inf', 'region_sup',\n",
       "       'region_nb_cpg', 'nb_cpg_found', 'nb_reads', 'global_cpg_fm',\n",
       "       'tot_nb_cpg', 'tot_nb_reads', 'cpg_fm', 'cpg_cov', 'cpg_pos',\n",
       "       'std_read_fm', 'mean_read_fm', 'std_cpg_fm', 'mean_cpg_fm',\n",
       "       'std_cpg_cov', 'mean_cpg_cov', 'std_cpg_dist', 'mean_cpg_dist',\n",
       "       'read_fm_kernel_0', 'read_fm_kernel_1', 'read_fm_kernel_2',\n",
       "       'read_fm_kernel_3', 'read_fm_kernel_4', 'read_fm_kernel_5',\n",
       "       'read_fm_kernel_6', 'read_fm_kernel_7', 'read_fm_kernel_8',\n",
       "       'read_fm_kernel_9', 'read_fm_kernel_10', 'cpg_fm_kernel_0',\n",
       "       'cpg_fm_kernel_1', 'cpg_fm_kernel_2', 'cpg_fm_kernel_3',\n",
       "       'cpg_fm_kernel_4', 'cpg_fm_kernel_5', 'cpg_fm_kernel_6',\n",
       "       'cpg_fm_kernel_7', 'cpg_fm_kernel_8', 'cpg_fm_kernel_9',\n",
       "       'cpg_fm_kernel_10', 'cpg_cov_kernel_0', 'cpg_cov_kernel_1',\n",
       "       'cpg_cov_kernel_2', 'cpg_cov_kernel_3', 'cpg_cov_kernel_4',\n",
       "       'cpg_dist_kernel_0', 'cpg_dist_kernel_1', 'cpg_dist_kernel_2',\n",
       "       'cpg_dist_kernel_3', 'cpg_dist_kernel_4', 'dnase_null', 'dnase_low',\n",
       "       'dnase_high', 'encode_ChiP_V2_null', 'encode_ChiP_V2_low',\n",
       "       'encode_ChiP_V2_high', 'tf_motifs_null', 'tf_motifs_low',\n",
       "       'tf_motifs_high'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef021547-c9fc-447c-93be-712ecf1d1b98",
   "metadata": {},
   "source": [
    "### Convert chromosome column in dummy variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "cf7ba911-d711-4864-ab2a-c9c66924e65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_df = pd.get_dummies(raw_df, columns = ['chr'], dtype=int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a837e1e-56f1-4d07-8e19-9e807f702b28",
   "metadata": {},
   "source": [
    "## Get rid of features that poorly correlate with ASM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1e1e3967-1cb0-4437-b2eb-25ac83e1acbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cpg_fm_kernel_0',\n",
       " 'read_fm_kernel_9',\n",
       " 'read_fm_kernel_10',\n",
       " 'read_fm_kernel_7',\n",
       " 'read_fm_kernel_6',\n",
       " 'cpg_fm_kernel_8',\n",
       " 'sample_category',\n",
       " 'cpg_fm_kernel_2',\n",
       " 'mean_cpg_dist',\n",
       " 'nb_reads',\n",
       " 'region_sup',\n",
       " 'region_inf',\n",
       " 'tf_motifs_high',\n",
       " 'tot_nb_reads',\n",
       " 'cpg_fm_kernel_6',\n",
       " 'tf_motifs_null',\n",
       " 'chr_14',\n",
       " 'cpg_fm_kernel_10',\n",
       " 'global_cpg_fm',\n",
       " 'cpg_fm_kernel_9',\n",
       " 'std_read_fm',\n",
       " 'chr_15',\n",
       " 'tf_motifs_low',\n",
       " 'mean_cpg_cov',\n",
       " 'cpg_fm_kernel_5',\n",
       " 'cpg_fm_kernel_3',\n",
       " 'cpg_fm_kernel_4',\n",
       " 'chr_13']"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_corr = raw_df.copy()\n",
    "corr_matrix = pd.DataFrame(abs(df_for_corr.corr()['asm_snp'])).sort_values(by = 'asm_snp')\n",
    "print(corr_matrix)\n",
    "\n",
    "model_scalar_variables = corr_matrix[corr_matrix['asm_snp'] > MIN_CORR].index.tolist()\n",
    "model_scalar_variables.remove('asm_snp')\n",
    "model_scalar_variables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ceeb4794-0de5-4320-8520-39ce44e16941",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nb_cpg_found</th>\n",
       "      <td>0.003222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_cov_kernel_4</th>\n",
       "      <td>0.003222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_dist_kernel_4</th>\n",
       "      <td>0.003948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr_19</th>\n",
       "      <td>0.004561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr_22</th>\n",
       "      <td>0.004561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr_13</th>\n",
       "      <td>0.156580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>asm_snp</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region_nb_cpg</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_cov_kernel_0</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_dist_kernel_0</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    asm_snp\n",
       "nb_cpg_found       0.003222\n",
       "cpg_cov_kernel_4   0.003222\n",
       "cpg_dist_kernel_4  0.003948\n",
       "chr_19             0.004561\n",
       "chr_22             0.004561\n",
       "...                     ...\n",
       "chr_13             0.156580\n",
       "asm_snp            1.000000\n",
       "region_nb_cpg           NaN\n",
       "cpg_cov_kernel_0        NaN\n",
       "cpg_dist_kernel_0       NaN\n",
       "\n",
       "[81 rows x 1 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "corr_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "60321953-6801-4c90-92bf-9c1701deb3fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>region_nb_cpg</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_cov_kernel_0</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_dist_kernel_0</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   asm_snp\n",
       "region_nb_cpg          NaN\n",
       "cpg_cov_kernel_0       NaN\n",
       "cpg_dist_kernel_0      NaN"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "corr_matrix[pd.isna(corr_matrix['asm_snp'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "f09cac27-a9ce-46f0-b6dd-2c9c86c667b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove features that poorly correlate with ASM\n",
    "model_variables_to_remove = corr_matrix[corr_matrix['asm_snp'] <= MIN_CORR].index.tolist()\n",
    "\n",
    "print(\"Removing\", len(model_variables_to_remove), \"variables from the datset\")\n",
    "\n",
    "for var in model_variables_to_remove:\n",
    "  raw_df.drop(var, axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "8d790c54-1480-45cc-abf1-5fbd4b45de3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove features where there is no correlation\n",
    "model_variables_to_remove = corr_matrix[pd.isna(corr_matrix['asm_snp'])].index.tolist()\n",
    "\n",
    "print(\"Removing\", len(model_variables_to_remove), \"variables from the datset\")\n",
    "\n",
    "for var in model_variables_to_remove:\n",
    "  raw_df.drop(var, axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d477209e-e129-4b3c-a2f0-5065da9a21bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Size of dataset:\", raw_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aae6224-d8e0-4444-8d19-6ef29d7742e1",
   "metadata": {},
   "source": [
    "## Create an \"image\" of the genomic region\n",
    "\n",
    "We create an 1D-image (length: genomic region interval) with 3 information per \"pixel\": CpG presence (0 or 1), CpG coverage, CpG fractional methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d30e5e77-f178-4d1b-b8b5-0d9754bd0855",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create arrays of positions, fractional methylation, and coverage for CpGs (~2h30)\n",
    "def create_genomic_array(df):\n",
    "  genomic_positions = []\n",
    "  genomic_fm = []\n",
    "  genomic_cov = [] \n",
    "  for position in range(df['region_inf'], df['region_sup'] + 1):\n",
    "    if position in df['cpg_pos']:\n",
    "      new_pos = 1\n",
    "      pos_index = df['cpg_pos'].index(position)\n",
    "      new_fm = df['cpg_fm'][pos_index]\n",
    "      new_cov = df['cpg_cov'][pos_index]\n",
    "    else:\n",
    "      new_pos = 0\n",
    "      new_fm = 0\n",
    "      new_cov = 0\n",
    "    genomic_positions = genomic_positions + [new_pos]\n",
    "    genomic_fm = genomic_fm + [new_fm]\n",
    "    genomic_cov = genomic_cov + [new_cov]\n",
    "  return np.transpose([genomic_positions, genomic_fm, genomic_cov])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "4ff4e0f2-9355-4174-8d14-cf1333af9495",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to the dataframe which we split before because it takes a lot of memory.\n",
    "\n",
    "# Split the dataframe into a list of dataframes of ~100k rows\n",
    "nb_dataframe_pieces = max(1, round(raw_df.shape[0]/100000))\n",
    "raw_df_pieces = np.array_split(raw_df, nb_dataframe_pieces)\n",
    "print(\"The dataframe has been split into\", nb_dataframe_pieces, \"pieces\")\n",
    "\n",
    "for df_piece in range(nb_dataframe_pieces):\n",
    "        print(\"processing the piece at position:\", df_piece)\n",
    "        raw_df_pieces[df_piece]['genomic_matrix'] = raw_df_pieces[df_piece].apply(lambda x: create_genomic_array(x), \n",
    "                                                                                      axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "d9db56c9-f255-4b35-a408-5346b6913593",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm_snp</th>\n",
       "      <th>sample_category</th>\n",
       "      <th>region_inf</th>\n",
       "      <th>region_sup</th>\n",
       "      <th>nb_reads</th>\n",
       "      <th>global_cpg_fm</th>\n",
       "      <th>tot_nb_reads</th>\n",
       "      <th>cpg_fm</th>\n",
       "      <th>cpg_cov</th>\n",
       "      <th>cpg_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_fm_kernel_8</th>\n",
       "      <th>cpg_fm_kernel_9</th>\n",
       "      <th>cpg_fm_kernel_10</th>\n",
       "      <th>tf_motifs_null</th>\n",
       "      <th>tf_motifs_low</th>\n",
       "      <th>tf_motifs_high</th>\n",
       "      <th>chr_13</th>\n",
       "      <th>chr_14</th>\n",
       "      <th>chr_15</th>\n",
       "      <th>genomic_matrix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167184001</td>\n",
       "      <td>167185000</td>\n",
       "      <td>71</td>\n",
       "      <td>0.815</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[0.742, 0.788, 0.25]</td>\n",
       "      <td>[31, 33, 40]</td>\n",
       "      <td>[167184497, 167184558, 167184725]</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4442</td>\n",
       "      <td>1.0919</td>\n",
       "      <td>0.1882</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>166743001</td>\n",
       "      <td>166744000</td>\n",
       "      <td>69</td>\n",
       "      <td>0.815</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[1, 0.9500000000000001, 0.929]</td>\n",
       "      <td>[38, 40, 42]</td>\n",
       "      <td>[166743591, 166743715, 166743733]</td>\n",
       "      <td>...</td>\n",
       "      <td>1.1904</td>\n",
       "      <td>3.2552</td>\n",
       "      <td>3.5369</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22376001</td>\n",
       "      <td>22377000</td>\n",
       "      <td>75</td>\n",
       "      <td>0.815</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[0.783, 0.838, 0.765]</td>\n",
       "      <td>[46, 37, 34]</td>\n",
       "      <td>[22376280, 22376368, 22376447]</td>\n",
       "      <td>...</td>\n",
       "      <td>3.7987</td>\n",
       "      <td>2.3026</td>\n",
       "      <td>0.5683</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63617001</td>\n",
       "      <td>63618000</td>\n",
       "      <td>57</td>\n",
       "      <td>0.815</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[0.636, 0.9570000000000001, 0.979]</td>\n",
       "      <td>[44, 47, 47]</td>\n",
       "      <td>[63617746, 63617751, 63617797]</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0022</td>\n",
       "      <td>2.1445</td>\n",
       "      <td>2.5149</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>188785001</td>\n",
       "      <td>188786000</td>\n",
       "      <td>144</td>\n",
       "      <td>0.815</td>\n",
       "      <td>307561222</td>\n",
       "      <td>[0.543, 0.356, 0.646]</td>\n",
       "      <td>[46, 90, 96]</td>\n",
       "      <td>[188785538, 188785705, 188785753]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4553</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>113881001</td>\n",
       "      <td>113882000</td>\n",
       "      <td>42</td>\n",
       "      <td>0.786</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[0.97, 0.765, 0.8290000000000001]</td>\n",
       "      <td>[33, 34, 35]</td>\n",
       "      <td>[113881397, 113881438, 113881453]</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8393</td>\n",
       "      <td>2.6090</td>\n",
       "      <td>1.6636</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>166743001</td>\n",
       "      <td>166744000</td>\n",
       "      <td>60</td>\n",
       "      <td>0.786</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[0.919, 0.962, 0.9490000000000001]</td>\n",
       "      <td>[37, 26, 39]</td>\n",
       "      <td>[166743591, 166743715, 166743733]</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4513</td>\n",
       "      <td>3.5827</td>\n",
       "      <td>3.3627</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>83870001</td>\n",
       "      <td>83871000</td>\n",
       "      <td>32</td>\n",
       "      <td>0.786</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[0.9520000000000001, 0.857, 0.909]</td>\n",
       "      <td>[21, 28, 22]</td>\n",
       "      <td>[83870175, 83870190, 83870236]</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2835</td>\n",
       "      <td>3.6985</td>\n",
       "      <td>2.5424</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>90796001</td>\n",
       "      <td>90797000</td>\n",
       "      <td>49</td>\n",
       "      <td>0.786</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[0.962, 0.722, 0.92]</td>\n",
       "      <td>[26, 36, 25]</td>\n",
       "      <td>[90796083, 90796160, 90796178]</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9863</td>\n",
       "      <td>2.6735</td>\n",
       "      <td>2.2307</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121825001</td>\n",
       "      <td>121826000</td>\n",
       "      <td>36</td>\n",
       "      <td>0.786</td>\n",
       "      <td>222365425</td>\n",
       "      <td>[0.788, 0.774, 0.742]</td>\n",
       "      <td>[33, 31, 31]</td>\n",
       "      <td>[121825759, 121825766, 121825777]</td>\n",
       "      <td>...</td>\n",
       "      <td>3.7298</td>\n",
       "      <td>1.6932</td>\n",
       "      <td>0.2917</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>985 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     asm_snp  sample_category  region_inf  region_sup  nb_reads  \\\n",
       "0          0                0   167184001   167185000        71   \n",
       "1          0                0   166743001   166744000        69   \n",
       "2          0                0    22376001    22377000        75   \n",
       "3          0                0    63617001    63618000        57   \n",
       "4          0                0   188785001   188786000       144   \n",
       "..       ...              ...         ...         ...       ...   \n",
       "995        0                0   113881001   113882000        42   \n",
       "996        0                0   166743001   166744000        60   \n",
       "997        0                0    83870001    83871000        32   \n",
       "998        0                0    90796001    90797000        49   \n",
       "999        0                0   121825001   121826000        36   \n",
       "\n",
       "     global_cpg_fm  tot_nb_reads                              cpg_fm  \\\n",
       "0            0.815     307561222                [0.742, 0.788, 0.25]   \n",
       "1            0.815     307561222      [1, 0.9500000000000001, 0.929]   \n",
       "2            0.815     307561222               [0.783, 0.838, 0.765]   \n",
       "3            0.815     307561222  [0.636, 0.9570000000000001, 0.979]   \n",
       "4            0.815     307561222               [0.543, 0.356, 0.646]   \n",
       "..             ...           ...                                 ...   \n",
       "995          0.786     222365425   [0.97, 0.765, 0.8290000000000001]   \n",
       "996          0.786     222365425  [0.919, 0.962, 0.9490000000000001]   \n",
       "997          0.786     222365425  [0.9520000000000001, 0.857, 0.909]   \n",
       "998          0.786     222365425                [0.962, 0.722, 0.92]   \n",
       "999          0.786     222365425               [0.788, 0.774, 0.742]   \n",
       "\n",
       "          cpg_cov                            cpg_pos  ...  cpg_fm_kernel_8  \\\n",
       "0    [31, 33, 40]  [167184497, 167184558, 167184725]  ...           2.4442   \n",
       "1    [38, 40, 42]  [166743591, 166743715, 166743733]  ...           1.1904   \n",
       "2    [46, 37, 34]     [22376280, 22376368, 22376447]  ...           3.7987   \n",
       "3    [44, 47, 47]     [63617746, 63617751, 63617797]  ...           1.0022   \n",
       "4    [46, 90, 96]  [188785538, 188785705, 188785753]  ...           0.4553   \n",
       "..            ...                                ...  ...              ...   \n",
       "995  [33, 34, 35]  [113881397, 113881438, 113881453]  ...           2.8393   \n",
       "996  [37, 26, 39]  [166743591, 166743715, 166743733]  ...           1.4513   \n",
       "997  [21, 28, 22]     [83870175, 83870190, 83870236]  ...           2.2835   \n",
       "998  [26, 36, 25]     [90796083, 90796160, 90796178]  ...           1.9863   \n",
       "999  [33, 31, 31]  [121825759, 121825766, 121825777]  ...           3.7298   \n",
       "\n",
       "     cpg_fm_kernel_9  cpg_fm_kernel_10  tf_motifs_null  tf_motifs_low  \\\n",
       "0             1.0919            0.1882               1              0   \n",
       "1             3.2552            3.5369               0              1   \n",
       "2             2.3026            0.5683               1              0   \n",
       "3             2.1445            2.5149               0              1   \n",
       "4             0.0551            0.0026               1              0   \n",
       "..               ...               ...             ...            ...   \n",
       "995           2.6090            1.6636               0              1   \n",
       "996           3.5827            3.3627               0              1   \n",
       "997           3.6985            2.5424               0              1   \n",
       "998           2.6735            2.2307               1              0   \n",
       "999           1.6932            0.2917               1              0   \n",
       "\n",
       "     tf_motifs_high  chr_13  chr_14  chr_15  \\\n",
       "0                 0       0       0       0   \n",
       "1                 0       0       0       0   \n",
       "2                 0       0       0       0   \n",
       "3                 0       0       1       0   \n",
       "4                 0       0       0       0   \n",
       "..              ...     ...     ...     ...   \n",
       "995               0       0       0       0   \n",
       "996               0       0       0       0   \n",
       "997               0       0       0       0   \n",
       "998               0       0       0       0   \n",
       "999               0       0       0       0   \n",
       "\n",
       "                                        genomic_matrix  \n",
       "0    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "1    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "2    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "3    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "4    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "..                                                 ...  \n",
       "995  [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "996  [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "997  [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "998  [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "999  [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...  \n",
       "\n",
       "[985 rows x 33 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_df_pieces[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a630899-e2ca-4037-bcd4-7a7d528b1bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete the variables we no longer need\n",
    "for var in ['cpg_cov', 'cpg_pos', 'cpg_fm', 'region_inf', 'region_sup']:\n",
    "      for df_piece in range(nb_dataframe_pieces):\n",
    "            print(\"processing the piece at position:\", df_piece)\n",
    "            raw_df_pieces[df_piece].drop(var, axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c23fd340-7827-4182-b371-554436af51f8",
   "metadata": {},
   "source": [
    "## Save dataframe with features on Cloud Storage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "9cf90a2f-dc29-4a84-8c3d-500e3bf49c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['asm_snp', 'sample_category', 'nb_reads', 'global_cpg_fm',\n",
       "       'tot_nb_reads', 'std_read_fm', 'mean_cpg_cov', 'mean_cpg_dist',\n",
       "       'read_fm_kernel_6', 'read_fm_kernel_7', 'read_fm_kernel_9',\n",
       "       'read_fm_kernel_10', 'cpg_fm_kernel_0', 'cpg_fm_kernel_2',\n",
       "       'cpg_fm_kernel_3', 'cpg_fm_kernel_4', 'cpg_fm_kernel_5',\n",
       "       'cpg_fm_kernel_6', 'cpg_fm_kernel_8', 'cpg_fm_kernel_9',\n",
       "       'cpg_fm_kernel_10', 'tf_motifs_null', 'tf_motifs_low', 'tf_motifs_high',\n",
       "       'chr_13', 'chr_14', 'chr_15', 'genomic_matrix'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the columns\n",
    "raw_df_pieces[0].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7cdade84-79fc-452e-9ab7-11317804c23d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-03-08_18-26-50\n"
     ]
    }
   ],
   "source": [
    "# Obtain the date/time\n",
    "now = datetime.today()\n",
    "dt_string = now.strftime(\"%Y-%m-%d_%H-%M-%S\")\n",
    "print(dt_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "52abf4ba-fd30-4102-b419-2713aeaa79f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the variable names to the bucket\n",
    "sys.stdout = open(\"variables.txt\", \"w\")\n",
    "%whos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "940d939f-8b48-4670-bba5-9877e03b3f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export variable names to to Cloud Storage\n",
    "!gsutil cp $var_file_name gs://$DEEPASM_BUCKET/notebook/$dt_string/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "67397ca8-78ff-44ac-8f7f-9114182fc3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "if EXPORT_PROCESSED_DATA == True:\n",
    "\n",
    "    for df_piece in range(nb_dataframe_pieces):\n",
    "        print(\"processing the piece at position:\", df_piece)\n",
    "        df_to_export = raw_df_pieces[df_piece]\n",
    "        print(\"Size of dataframe:\", df_to_export.shape)\n",
    "\n",
    "        print(\"Saving the file as HDF5...\")\n",
    "        file_name = \"prepared_df_\" + str(df_piece) + \".h5\"\n",
    "        print(\"File name:\", file_name)\n",
    "        df_to_export.to_hdf(file_name, key = 'df', mode = 'w')\n",
    "\n",
    "        print(\"Exporting file to bucket...\")\n",
    "        !gsutil cp $file_name gs://$DEEPASM_BUCKET/notebook/$dt_string/\n",
    "else:\n",
    "        print(\"Not exporting the scaled DF per variable\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6636177-ca5d-4c77-845c-71467c3755e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Importing prepared features from bucket"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "tf2-gpu.2-7.m87",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/tf2-gpu.2-7:m87"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
